File: Data-Structure/Nodejs/LeetCode by Category/LinkedList/143. Reorder List.js
//143. Reorder List
/*
You are given the head of a singly linked-list. The list can be represented as:

L0 → L1 → … → Ln - 1 → Ln
Reorder the list to be on the following form:

L0 → Ln → L1 → Ln - 1 → L2 → Ln - 2 → …
You may not modify the values in the list's nodes. Only nodes themselves may be changed.
*/
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
 var reorderList = function(head) { 
    if (head === null || head.next === null || head.next.next === null) return head;
    
    let dummy = new ListNode(0);
    let p = dummy;
    let p1 = head;
    
    let fast = head;
    let slow = head;
    let preSlow = head;
    
    // find middle
    while (fast && fast.next) {
        fast = fast.next.next;
        preSlow = slow;
        slow = slow.next;
    }
    
    let p2 = slow;
    let curr = slow;
    let pre = null;
    let next = null;

    // reverse
    while (curr) {
        next = curr.next;
        curr.next = pre;
        pre = curr;
        curr = next
    } 
    p2 = pre; 
    
    preSlow.next  = null;
    
    // merge
    while (p1 && p2) {
        p.next = p1;
        p1 = p1.next;
        p = p.next;
        
        p.next = p2;
        p2 = p2.next;
        p = p.next;
    }
    
    return dummy.next;
    
};

/*
Convert Sorted List to Binary Search Tree
Swapping Nodes in a Linked List
Convert Binary Number in a Linked List to Integer
*/

File: Data-Structure/_All-In-One-LeetCode/00143. Reorder List/143. Reorder List.py
# 143. Reorder List
'''
You are given the head of a singly linked-list. The list can be represented as:

L0 → L1 → … → Ln - 1 → Ln
Reorder the list to be on the following form:

L0 → Ln → L1 → Ln - 1 → L2 → Ln - 2 → …
You may not modify the values in the list's nodes. Only nodes themselves may be changed.

'''


from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverse(self, head):
        if head is None:
            return head
        if head.next is None:
            return head
        
        curr = head
        pre, nxt = None, None
        while curr:
            nxt = curr.next
            curr.next = pre
            pre = curr
            curr = nxt
        
        return pre



    def reorder_list(self, head: Optional[ListNode]) -> None:
        if head is None:
            return head
        if head.next is None:
            return head

        slow, fast = head, head
        # revert second half

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        second = slow.next
        slow.next = None

        second = self.reverse(second)

        # merge tow half
        first = head
        while second:
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first, second = tmp1, tmp2





